libc/vga/writev-vga.c

```diff
@@ -0,0 +1,129 @@
+│ binary, for any purpose, commercial or non-commercial, and by any            │
+│ means.                                                                       │
+│                                                                              │
+│ In jurisdictions that recognize copyright laws, the author or authors        │
+│ of this software dedicate any and all copyright interest in the              │
+│ software to the public domain. We make this dedication for the benefit       │
+│ of the public at large and to the detriment of our heirs and                 │
+│ successors. We intend this dedication to be an overt act of                  │
+│ relinquishment in perpetuity of all present and future rights to this        │
+│ software under copyright law.                                                │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,              │
+│ EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF           │
+│ MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.       │
+│ IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR            │
+│ OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,        │
+│ ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR        │
+│ OTHER DEALINGS IN THE SOFTWARE.                                              │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/calls/struct/fd.internal.h"
+#include "libc/calls/struct/iovec.h"
+#include "libc/calls/struct/iovec.internal.h"
+#include "libc/vga/vga.h"
+#include "libc/runtime/pc.internal.h"
+#include "libc/str/str.h"
+
+#define CRTPORT 0x3d4
+#define WIDTH 80
+
+typedef struct {
+  char ch;
+  uint8_t attr;
+} char_cell_t;
+
+typedef char_cell_t char_row_t[WIDTH];
+
+const char vga_console[0];
+
+static unsigned short height = 25, curr_row = 12, curr_col = 0;
+static uint8_t curr_attr = 0x07;
+
+static void scroll(void) {
+  unsigned j;
+  uint8_t attr = curr_attr;
+  char_row_t * const vid_buf = (char_row_t *)(BANE + 0xb8000ull);
+  memmove(vid_buf, vid_buf + 1, (height - 1) * sizeof(char_row_t));
+  for (j = 0; j < WIDTH; ++j)
+    vid_buf[height - 1][j] = (char_cell_t){ ' ', attr };
+}
+
+static void may_scroll(void) {
+  if (curr_col >= WIDTH) {
+    curr_col = 0;
+    scroll();
+  }
+}
+
+static void updatexy_vga(void) {
+  unsigned short pos = curr_row * WIDTH + curr_col;
+  outb(CRTPORT, 0x0e);
+  outb(CRTPORT + 1, (unsigned char)(pos >> 8));
+  outb(CRTPORT, 0x0f);
+  outb(CRTPORT + 1, (unsigned char)pos);
+}
+
+static void writec_vga(char c) {
+  /* TODO: ensure screen in a known mode (text or graphics), at rlinit time */
+  /* TODO: use our own font, rather than rely on BIOS's CP437 font */
+  /* TODO: handle UTF-8 multi-bytes */
+  /* TODO: handle VT102 escape sequences */
```

Comment:
Is https://github.com/jart/cosmopolitan/blob/master/tool/build/lib/pty.c something that might be useful to you? It's how Blinkenlights displays a terminal inside a terminal.

Reply:
> Is https://github.com/jart/cosmopolitan/blob/master/tool/build/lib/pty.c something that might be useful to you?

Thank you &mdash; yes, this looks like it might be useful to me.  Let me take a further look at the `pty.c` code and see how I can use it.